MathLib/LinAlg/MatrixVectorTraits.h

```diff
@@ -0,0 +1,72 @@
+/**
+ * \copyright
+ * Copyright (c) 2012-2016, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+
+#ifndef MATHLIB_MATRIX_VECTOR_TRAITS_H
+#define MATHLIB_MATRIX_VECTOR_TRAITS_H
+
+#include<memory>
+
+#include "MatrixProviderUser.h"
+
+namespace MathLib
+{
+template<typename Matrix>
+struct MatrixVectorTraits;
+}
+
+#define SPECIALIZE_MATRIX_VECTOR_TRAITS(MATVEC, IDX) \
+    template<> struct MatrixVectorTraits<MATVEC> { \
+        using Index = IDX; \
+        static std::unique_ptr<MATVEC> newInstance(); \
+        static std::unique_ptr<MATVEC> newInstance(MATVEC const& A); \
+        static std::unique_ptr<MATVEC> newInstance(MatrixSpecifications const& spec); \
+    };
+
+
+#ifdef OGS_USE_EIGEN
+
+#include<Eigen/Core>
+
+namespace MathLib
+{
+SPECIALIZE_MATRIX_VECTOR_TRAITS(Eigen::MatrixXd, Eigen::MatrixXd::Index);
+SPECIALIZE_MATRIX_VECTOR_TRAITS(Eigen::VectorXd, Eigen::VectorXd::Index);
+}
+
+#endif
+
+
+#ifdef USE_PETSC
+
+#include "MathLib/LinAlg/PETSc/PETScMatrix.h"
+#include "MathLib/LinAlg/PETSc/PETScVector.h"
+
+namespace MathLib
+{
+SPECIALIZE_MATRIX_VECTOR_TRAITS(PETScMatrix, PETScMatrix::IndexType);
+SPECIALIZE_MATRIX_VECTOR_TRAITS(PETScVector, PETScVector::IndexType);
+}
+
+
+#elif defined(OGS_USE_EIGEN)
```

Comment:
Same as in the implementation: Why not merge the two OGS_USE_EIGEN sections into one?